zebra-chain/src/sapling/tree.rs

```diff
@@ -376,3 +423,108 @@ impl From<Vec<jubjub::Fq>> for NoteCommitmentTree {
+/// consistent with `zcashd` in RPCs. Note that [`NoteCommitmentTree`] itself is
+/// represented as [`Frontier`](bridgetree::Frontier).
+///
+/// The formats are semantically equivalent. The primary difference between them
+/// is that in [`Frontier`](bridgetree::Frontier), the vector of parents is
+/// dense (we know where the gaps are from the position of the leaf in the
+/// overall tree); whereas in [`CommitmentTree`](merkle_tree::CommitmentTree),
+/// the vector of parent hashes is sparse with [`None`] values in the gaps.
+///
+/// The sparse format, used in this implementation, allows representing invalid
+/// commitment trees while the dense format allows representing only valid
+/// commitment trees.
+///
+/// It is likely that the dense format will be used in future RPCs, in which
+/// case the current implementation will have to change and use the format
+/// compatible with [`Frontier`](bridgetree::Frontier) instead.
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize)]
+pub struct SerializedTree(Vec<u8>);
+
+impl From<&NoteCommitmentTree> for SerializedTree {
+    fn from(tree: &NoteCommitmentTree) -> Self {
+        let mut serialized_tree = vec![];
+
+        // Convert the note commitment tree represented as a frontier into the
+        // format compatible with `zcashd`.
+        //
+        // `librustzcash` has a function [`from_frontier()`][1], which returns a
+        // commitment tree in the sparse format. However, the returned tree
+        // always contains [`MERKLE_DEPTH`] parent nodes, even though some
+        // trailing parents are empty. Such trees are incompatible with Sapling
+        // commitment trees returned by `zcashd` because `zcashd` returns
+        // Sapling commitment trees without empty trailing parents. For this
+        // reason, Zebra implements its own conversion between the dense and
+        // sparse formats for Sapling.
+        //
+        // [1]: <https://github.com/zcash/librustzcash/blob/a63a37a/zcash_primitives/src/merkle_tree.rs#L125>
+        if let Some(frontier) = tree.inner.value() {
+            let (left_leaf, right_leaf) = match frontier.leaf() {
+                Leaf::Left(left_value) => (Some(left_value), None),
+                Leaf::Right(left_value, right_value) => (Some(left_value), Some(right_value)),
+            };
+
+            let mut ommers_iter = frontier.ommers().iter();
+            let mut position: usize = frontier.position().into();
+            let mut parents = vec![];
+
+            position &= !1;
+
+            for i in 1..MERKLE_DEPTH {
+                let bit_mask = 1 << i;
+
+                if position & bit_mask == 0 {
+                    parents.push(None);
+                } else {
+                    parents.push(ommers_iter.next());
+                    position &= !bit_mask;
+                    if position == 0 {
+                        break;
+                    }
+                }
+            }
```

Comment:
That was a neat way to solve it!

But it would be good to have some comments to help whoever is reading the code. Ideally we'd also to explain the original code; but I just explaining the difference between the original zcashd code is enough for now. (I think it would be better to comment the original code instead, we can consider that later)

Here's a suggestion, feel free to tweak it (or fix it if I got something wrong).

```suggestion
           // Run through the bits of `position` and push a "ommer" for each bit `1` (or `None` otherwise), 
           // skipping the lower bit (which indicates the leaf).
           // In contrast to the zcashd code linked above, we want to skip any trailing `None` parents 
           // at the top of the tree. To do that, we clear the bits as we go through them and
           // break early if the remaining bits are all zero (i.e. `position` is zero).

            // Clear the lower bit of the position since we'll skip it
            position &= !1;

            for i in 1..MERKLE_DEPTH {
                let bit_mask = 1 << i;

                if position & bit_mask == 0 {
                    parents.push(None);
                } else {
                    parents.push(ommers_iter.next());
                    // Clear the bit we just read
                    position &= !bit_mask;
                    // The rest of the bits are all zero; exit early
                    if position == 0 {
                        break;
                    }
                }
            }
```

Reply:
Sure, the description is great. I'll commit it, tweak it a bit locally, and then push back here.